lib/utils/prompt/context_reader_test.go

```diff
@@ -186,6 +186,70 @@ func TestContextReader_ReadPassword(t *testing.T) {
 	})
 }
 
+func TestNotifyExit_restoresTerminal(t *testing.T) {
+	oldStdin := Stdin()
+	t.Cleanup(func() { SetStdin(oldStdin) })
+
+	pr, _ := io.Pipe()
+
+	devNull, err := os.OpenFile(os.DevNull, os.O_RDWR, 0666)
+	require.NoError(t, err, "Failed to open %v", os.DevNull)
+	defer devNull.Close()
+
+	term := &fakeTerm{reader: pr}
+	ctx := context.Background()
+
+	tests := []struct {
+		name        string
+		doRead      func(ctx context.Context, cr *ContextReader) error
+		wantRestore bool
+	}{
+		{
+			name: "no pending read",
+			doRead: func(ctx context.Context, cr *ContextReader) error {
+				<-ctx.Done()
+				return ctx.Err()
+			},
+		},
+		{
+			name: "pending clean read",
+			doRead: func(ctx context.Context, cr *ContextReader) error {
+				_, err := cr.ReadContext(ctx)
+				return err
+			},
+		},
+		{
+			name: "pending password read",
+			doRead: func(ctx context.Context, cr *ContextReader) error {
+				_, err := cr.ReadPassword(ctx)
+				return err
+			},
+			wantRestore: true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			term.restoreCalled = false // reset state between tests
+
+			cr := NewContextReader(pr)
+			cr.term = term
+			cr.fd = int(devNull.Fd()) // arbitrary
+			SetStdin(cr)
+
+			ctx, cancel := context.WithCancel(ctx)
+			go func() {
+				time.Sleep(1 * time.Millisecond) // give the read time to block
+				cancel()
+			}()
+			err := test.doRead(ctx, cr)
+			require.ErrorIs(t, err, context.Canceled, "unexpected read error")
+
+			NotifyExit() // closes Stdin
```

Comment:
Really pedantic nitpick, but currently that `defer devNull.Close()` is load-bearing.

```suggestion
			NotifyExit() // closes Stdin
			runtime.KeepAlive(devNull)
```

Reply:
I'm not sure I follow, the defer is in the outer function.

Reply:
Yes, we're relying on the defer being there to prevent the finalizer from closing `devNull` before we're done with the file descriptor - it feels a little untidy, even though it's perfectly fine as it's coded.

Reply:
IOWs, I think it's good practice for every call to `(*os.File).Fd()` to be paired with a `runtime.KeepAlive` once you're done with the fd.